File: MyGame/app/src/main/java/com/example/mygame/object/bonuses/Spring.kt
package com.example.mygame.`object`.bonuses

import android.animation.ValueAnimator
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import com.example.mygame.R
import com.example.mygame.`interface`.IBonus
import com.example.mygame.`interface`.IDrawable
import com.example.mygame.`interface`.IGameObject
import com.example.mygame.`interface`.IMoveable
import com.example.mygame.`interface`.IVisitor
import com.example.mygame.`object`.Platform
import com.example.mygame.`object`.Player

class Spring(private val initBitmap: MutableList<Bitmap>,
             createdX: Float,
             createdY: Float
) : IDrawable, IMoveable, IBonus, IGameObject {
    private val animationDuration : Long = 150

    private var currentFrame = 0
    private var bitmap = initBitmap[currentFrame]
    private var isStretchRunning = false
    private var animator : ValueAnimator? = null

    override var x = createdX
    override var y = createdY

    override var left = x - WIDTH / 2
    override var right = x + WIDTH / 2
    override var top = y - WIDTH / 2
    override var bottom = y + WIDTH / 2

    override var isDisappeared = false

    fun runStretchAnimation() {
        if (!isStretchRunning) {
            isStretchRunning = true
            animator = ValueAnimator.ofInt(0, initBitmap.size - 1).apply {
                this.duration = animationDuration
                addUpdateListener { animator ->
                    currentFrame = animator.animatedValue as Int
                    bitmap = initBitmap[currentFrame]
                    setPosition(x, y - ANIMATION_HEIGHT_VALUE)
                }
            }
            animator?.start()
        }
    }

    override fun draw(canvas: Canvas) {
        canvas.drawBitmap(bitmap, left, top, null)
    }

    override fun setPosition(startX: Float, startY: Float) {
        x = startX
        y = startY
        left = x - WIDTH / 2
        right = x + WIDTH / 2
        top = y - HEIGHT / 2
        bottom = y + HEIGHT /2
    }

    override fun updatePositionX(newX: Float) {}
    override fun updatePositionY(elapsedTime: Float) {}

    override fun accept(visitor: IVisitor) {
        visitor.visit(this)
    }

    companion object {
        private const val WIDTH = 78f
        private const val HEIGHT = 53f
        private const val ANIMATION_HEIGHT_VALUE = 5f
    }
}

File: GameServer/src/Model/Bonus/Spring.php
<?php

namespace App\Model\Bonus;

use App\Model\Platform;
use App\Model\IMoveable;
use App\Model\IGameObject;
use App\Collision\IVisitor;
use App\Constants\JsonConstants;

class Spring implements IMoveable, IBonus, IGameObject
{
    private const WIDTH = 78.0;
    private const HEIGHT = 53.0;

    private float $x;
    private float $y;

    private int $id;

    private float $left;
    private float $right;
    private float $top;
    private float $bottom;

    private bool $isDisappeared = false;

    public function __construct(Platform $platform, $id)
    {
        $this->id = $id;

        $this->x = $platform->getX();
        $this->y = $platform->getTop() - self::HEIGHT / 2;

        $this->left = $this->x - self::WIDTH / 2;
        $this->right = $this->x + self::WIDTH / 2;
        $this->top = $this->y - self::HEIGHT / 2;
        $this->bottom = $this->y + self::HEIGHT / 2;
    }

    public function setPosition(float $startX, float $startY): void
    {
        $this->x = $startX;
        $this->y = $startY;

        $this->left = $this->x - self::WIDTH / 2;
        $this->right = $this->x + self::WIDTH / 2;
        $this->top = $this->y - self::HEIGHT / 2;
        $this->bottom = $this->y + self::HEIGHT / 2;
    }

    public function accept(IVisitor $visitor): void
    {
        $visitor->visitSpring($this);
    }

    public function getX(): float
    {
        return $this->x;
    }

    public function getY(): float
    {
        return $this->y;
    }

    public function getLeft(): float
    {
        return $this->left;
    }

    public function getRight(): float
    {
        return $this->right;
    }

    public function getTop(): float
    {
        return $this->top;
    }

    public function getBottom(): float
    {
        return $this->bottom;
    }

    public function getIsDisappeared(): bool
    {
        return $this->isDisappeared;
    }

    public function setIsDisappeared(bool $value): void
    {
        $this->isDisappeared = $value;
    }

    public function updatePosition(float $elapsedTime): void
    {
    }

    public function toArray(): array
    {
        return [
            JsonConstants::SPRING_BONUS_TYPE,
            $this->x,
            $this->y,
            $this->id,
            0,
        ];
    }
}
